brave/brave-core
# Enable granular control of autoplay

by following steps:
1. Add AutoplayPermissionContext
2. Update PermissionDescriptorToPermissionType
3. RquestPermission from AutoplayPolicy
4. Content setting page & blocked status bubble
5. Auto refresh when clicking allow in permission ask bubble

fix https://github.com/brave/brave-browser/issues/281

Auditors: @bbondy, @bridiver 

## Submitter Checklist:


## Test Plan:
Automatic:
```
yarn test brave_unit_tests --filter=AutoplayPermissionContextTests.*
yarn test brave_browser_tests --filter=AutoplayPermissionContextBrowserTest.*
yarn test brave_browser_tests --filter=BraveAutoplayBlockedImageModelTest.*
```
Manual:
a. **Allow autoplay**

1. Go to https://www.cnet.com/ pick a random article
2. There should be a bubble shows up asking allow/block www.cnet.com to autoplay media <img width="326" alt="screen shot 2018-07-09 at 6 51 03 pm" src="https://user-images.githubusercontent.com/11330831/42484688-12da1348-83a9-11e8-9a49-6a13cad6e6b1.png">
 3. Click allow and the page should auto refresh and video should play
 4. Check Autoplay site setting should be allow
 4.a In site detail <img width="325" alt="screen shot 2018-07-09 at 6 51 53 pm" src="https://user-images.githubusercontent.com/11330831/42484761-602a6b70-83a9-11e8-8a34-5a26d9ef4407.png">
 4.b In site content setting <img width="713" alt="screen shot 2018-07-09 at 6 55 26 pm" src="https://user-images.githubusercontent.com/11330831/42484824-acb6ee14-83a9-11e8-9230-ddbd5f08c7b8.png">
 4.c In global content setting <img width="750" alt="screen shot 2018-07-09 at 6 57 08 pm" src="https://user-images.githubusercontent.com/11330831/42484887-f0024c04-83a9-11e8-8709-5056516be32a.png">
 5. Next time you visit cnet.com, the video will be autoplay automatically


b. **Block autoplay**
1. Same steps as a.
2. Except click block will not refresh the page
3. Notice when video is blocked there will be an indicator on urlbar 
<img width="360" alt="screen shot 2018-07-09 at 7 12 05 pm" src="https://user-images.githubusercontent.com/11330831/42485324-006a7a1a-83ac-11e8-9063-de77bf7c815e.png">


c. **General behavior**
1. every time you change site content setting, there will be a banner shows up 
<img width="1196" alt="screen shot 2018-07-09 at 7 12 39 pm" src="https://user-images.githubusercontent.com/11330831/42485349-2364da56-83ac-11e8-8e5d-6c7ec9260dff.png">
2. If you close permission bubble, that would be automatically blocked <img width="350" alt="screen shot 2018-07-09 at 7 15 33 pm" src="https://user-images.githubusercontent.com/11330831/42485446-7b92b07c-83ac-11e8-902f-182a64d834d8.png">
<img width="326" alt="screen shot 2018-07-09 at 7 16 21 pm" src="https://user-images.githubusercontent.com/11330831/42485477-9b7dab3a-83ac-11e8-8dd1-323508242066.png">
3. Open any local media files should allow autoplay by default


## Reviewer Checklist:
